### DISCUSSION TITLE
"yarn install" fails on Windows

### DISCUSSION THREAD

**Discussion by CannyDuck** [eclipse-theia/theia#8282] [General] [answered]
If i try to run "yarn install" it hangs during installation. Any idea?

```
[3/4] Linking dependencies...
warning "@theia/callhierarchy > @theia/monaco > onigasm > tslint@5.20.1" has unmet peer dependency "typescript@>=2.3.0-dev || >=2.4.0-dev || >=2.5.0-dev || >=2.6.0-dev ||
 >=2.7.0-dev || >=2.8.0-dev || >=2.9.0-dev || >=3.0.0-dev || >= 3.1.0-dev || >= 3.2.0-dev".
warning "@theia/callhierarchy > @theia/monaco > onigasm > tslint > tsutils@2.29.0" has unmet peer dependency "typescript@>=2.1.0 || >=2.1.0-dev || >=2.2.0-dev || >=2.3.0-
dev || >=2.4.0-dev || >=2.5.0-dev || >=2.6.0-dev || >=2.7.0-dev || >=2.8.0-dev || >=2.9.0-dev || >= 3.0.0-dev || >= 3.1.0-dev".
warning "@theia/git > @theia/scm > react-autosize-textarea@7.1.0" has unmet peer dependency "react@^0.14.0 || ^15.0.0 || ^16.0.0".
warning "@theia/git > @theia/scm > react-autosize-textarea@7.1.0" has unmet peer dependency "react-dom@^0.14.0 || ^15.0.0 || ^16.0.0".
warning "@theia/messages > react-perfect-scrollbar@1.5.8" has unmet peer dependency "react@>=16.3.3".
warning "@theia/messages > react-perfect-scrollbar@1.5.8" has unmet peer dependency "react-dom@>=16.3.3".
warning "@theia/cli > @theia/application-manager > font-awesome-webpack@0.0.5-beta.2" has unmet peer dependency "font-awesome@>=4.3.0".
warning "@theia/cli > @theia/application-manager > font-awesome-webpack > less-loader@2.2.3" has incorrect peer dependency "less@^2.3.1".
warning Workspaces can only be enabled in private projects.
warning Workspaces can only be enabled in private projects.
[4/4] Building fresh packages...
[-/12] ? waiting...
[6/12] ? vscode-ripgrep
[-/12] ? waiting...
[-/12] ? waiting...
[-/12] ? waiting...
```

**Comment by benoitf** [comment]
@CannyDuck do you have latest yarn version ?

Also is that your internet connectivity is "high" or "low" ?

**Comment by CannyDuck** [comment]
I use version 1.22.4. I configured a proxy for yarn, other dependencies are also successful retrieved during the install process.

**Comment by benoitf** [comment]
try with `--verbose`  flag on yarn command line to see where it fails ?

**Reply by CannyDuck** [reply]
Already, tried and don't get any further details.

**Comment by MikaelDesharnais-ST** [comment]
There is a bug in the vscode-ripgrep extension. It does not support proxies. There is an issue on this point but still no PR : https://github.com/eclipse-theia/theia/issues/8171
On our systems (which require a proxy too), it crashes but maybe in your case it just hangs. You can try the following 

- Kill the yarn command
- Go to node_modules/vscode-ripgrep
- remove the bin directory
- Install global-agent package : npm i global-agent
- Create an environment variable GLOBAL_AGENT_HTTP_PROXY which contains yours proxy configuration
- go in the lib directory
- run this command :  node -r 'global-agent/bootstrap' postinstall.js
- go back to the root of your folder & restart yarn

hope this helps

**Reply by CannyDuck** [reply]
Thanks for the workaroud. Which lib folder I have to choose "go in the lib directory"? I tried the theia lib folder and getting this error.

```
D:\projects\theia\lib> node -r 'global-agent/bootstrap' postinstall.js
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module ''global-agent/bootstrap''
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at Module._preloadModules (internal/modules/cjs/loader.js:901:12)
    at preloadModules (internal/bootstrap/node.js:602:7)
    at startup (internal/bootstrap/node.js:273:9)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
```

**Comment by vince-fugnitto** [comment]
The pull-request to address the issue is here: https://github.com/eclipse-theia/theia/pull/8280
If the current `vscode-ripgrep` causes problems (due to proxy settings), you can add a `yarn resolution` to pin it to a correct version.

**Comment by MikaelDesharnais-ST** [comment]
The lib directory is the node_modules/vscode-ripgrep/lib

**Reply by CannyDuck** [reply]
```
D:\projects\theia\node_modules\vscode-debugprotocol\lib> node -r 'global-agent/bootstrap' postinstall.js
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module ''global-agent/bootstrap''
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at Module._preloadModules (internal/modules/cjs/loader.js:901:12)
    at preloadModules (internal/bootstrap/node.js:602:7)
    at startup (internal/bootstrap/node.js:273:9)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
```

**Comment by MikaelDesharnais-ST** [comment]
Are you sure that you did npm i global-agent and that it installed global-agent properly ?

**Reply by CannyDuck** [reply]
Yes, may boostrap plays a role here?

```
D:\projects\theia\node_modules\vscode-debugprotocol\lib>npm i global-agent

> core-js@3.6.5 postinstall D:\Projekte\theia\node_modules\vscode-debugprotocol\node_modules\core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js on Open Collective or Patreon:
> https://opencollective.com/core-js
> https://www.patreon.com/zloirock

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)

npm notice created a lockfile as package-lock.json. You should commit this file.
+ global-agent@2.1.12
added 17 packages from 19 contributors and audited 17 packages in 3.738s

5 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities


D:\projects\theia\node_modules\vscode-debugprotocol\lib> node -r 'global-agent/bootstrap' postinstall.js
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module ''global-agent/bootstrap''
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at Module._preloadModules (internal/modules/cjs/loader.js:901:12)
    at preloadModules (internal/bootstrap/node.js:602:7)
    at startup (internal/bootstrap/node.js:273:9)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
```

**Comment by MikaelDesharnais-ST** [comment]
You are in the wrong folder : vscode-debugprotocol instead of vscode-ripgrep

**Reply by CannyDuck** [reply]
Sorry, but same result.

```
D:\projects\theia\node_modules\vscode-ripgrep> node -r 'global-agent/bootstrap' postinstall.js
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module ''global-agent/bootstrap''
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at Module._preloadModules (internal/modules/cjs/loader.js:901:12)
    at preloadModules (internal/bootstrap/node.js:602:7)
    at startup (internal/bootstrap/node.js:273:9)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
```

**Comment by MikaelDesharnais-ST** [comment]
Did you install global-agent in the vscode-ripgrep folder ?
Just to be sure : 
npm i global-agent && cd lib && node -r 'global-agent/bootstrap' postinstall.js 
in the vscode-ripgrep folder.

**Reply by CannyDuck** [reply]
Yes

´´´
D:\projects\theia\node_modules\vscode-ripgrep>npm i global-agent
+ global-agent@2.1.12
updated 1 package and audited 17 packages in 0.683s
found 0 vulnerabilities


D:\projects\theia\node_modules\vscode-ripgrep>node -r 'global-agent/bootstrap' postinstall.js
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module ''global-agent/bootstrap''
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at Module._preloadModules (internal/modules/cjs/loader.js:901:12)
    at preloadModules (internal/bootstrap/node.js:602:7)
    at startup (internal/bootstrap/node.js:273:9)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
´´´

**Comment by vince-fugnitto** [comment, marked as answer]
I'm not sure why all this global extra agent stuff is required, the simplest would just be to include a `resolution` in the root `package.json` and rebuild:

```json
"resolutions": {
    "**/vscode-ripgrep": "1.8.0"
},
```

- https://github.com/eclipse-theia/theia/discussions/8282#discussioncomment-44978

**Reply by MikaelDesharnais-ST** [reply]
It is required because the module completly ignores the proxy, of course, your solution is better in the long term.

**Reply by vince-fugnitto** [reply]
This solution is also much simpler than having to do the workaround.
The fix is landing soon so a `resolution` would be the best approach for now.

**Reply by CannyDuck** [reply]
I tried it like this, but still hangs.

```json
{
  "private": true,
  "dependencies": {
    "@theia/callhierarchy": "next",
    "@theia/file-search": "next",
    "@theia/git": "next",
    "@theia/markers": "next",
    "@theia/messages": "next",
    "@theia/mini-browser": "next",
    "@theia/navigator": "next",
    "@theia/outline-view": "next",
    "@theia/plugin-ext-vscode": "next",
    "@theia/preferences": "next",
    "@theia/preview": "next",
    "@theia/search-in-workspace": "next",
    "@theia/terminal": "next",
    "@theia/java": "next"
  },
  "devDependencies": {
    "@theia/cli": "next"
  },
  "resolutions": {
    "**/vscode-ripgrep": "1.8.0"
  }
}
```

**Reply by vince-fugnitto** [reply]
@CannyDuck if you have a `yarn.lock` it will need to be updated as to not pin the dependency, can you confirm if you have one?

**Reply by CannyDuck** [reply]
I have no `yarn.lock`

**Reply by vince-fugnitto** [reply]
> I have no `yarn.lock`

In your `node_modules/vscode-ripgrep` do you have `vscode-ripgrep` as `1.8.0` when viewing it's package.json?
For example, in my case I have:

```json
{
  "name": "vscode-ripgrep",
  "version": "1.8.0",
  "description": "A module for using ripgrep in a Node project",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/microsoft/vscode-ripgrep"
  },
  "scripts": {
    "postinstall": "node ./lib/postinstall.js"
  },
  "author": "Rob Lourens",
  "license": "MIT",
  "dependencies": {
    "https-proxy-agent": "^4.0.0",
    "proxy-from-env": "^1.1.0"
  },
  "devDependencies": {
    "@types/node": "^10.12.0"
  }
}
```

**Reply by CannyDuck** [reply]
Yes, my file look like yours.

**Reply by vince-fugnitto** [reply]
> Yes, my file look like yours.

Unfortunately I cannot identify the problem (it might not be related entirely to `vscode-ripgrep` after all).
Were you ever successful in building the application before?

For example, we have [prerequisites](https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#prerequisites), and building steps specifically for [Windows](https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#building-on-windows).

**Reply by CannyDuck** [reply]
I make a fresh installation with the steps described in the Windows instruction. Now it works. Thanks a lot.

**Comment by MikaelDesharnais-ST** [comment]
That's very strange, could you check what's in your vscode-ripgrep/node_modules :
dir D:\projects\theia\node_modules\vscode-ripgrep\node_modules

**Reply by CannyDuck** [reply]
folder: .vscode
folder: bin
folder: lib
jsconfig.json
LICENSE
package.json
README.md
